harness/tests/leader_election.rs

```diff
@@ -1,120 +1,68 @@
-use std::{sync::mpsc, time::Duration};
-use tokio::time::{sleep, timeout};
+use raftify::{raft::StateRole, StableStorage};
+use std::sync::mpsc;
 
 use harness::{
     constant::{FIVE_NODE_EXAMPLE, THREE_NODE_EXAMPLE},
     raft::{build_raft_cluster, wait_until_rafts_ready, Raft},
     utils::{
-        kill_previous_raft_processes, load_peers, wait_for_until_cluster_size_decrease,
-        wait_for_until_cluster_size_increase,
+        collect_state_matching_rafts_until_leader_emerge, kill_previous_raft_processes, load_peers,
     },
 };
 
-#[tokio::test]
-pub async fn test_leader_election_in_three_node_example() {
+async fn run_leader_election_test(example_filename: &str, mut raft_list: Vec<u64>) {
     kill_previous_raft_processes();
 
     let (tx_raft, rx_raft) = mpsc::channel::<(u64, Raft)>();
-
-    let peers = load_peers(THREE_NODE_EXAMPLE).await.unwrap();
+    let peers = load_peers(example_filename).await.unwrap();
     let _raft_tasks = tokio::spawn(build_raft_cluster(tx_raft, peers.clone()));
-    sleep(Duration::from_secs(1)).await;
-
-    let mut rafts = wait_until_rafts_ready(None, rx_raft, 3).await;
-
-    let raft_1 = rafts.get_mut(&1).unwrap();
-
-    wait_for_until_cluster_size_increase(raft_1.clone(), 3).await;
-
-    sleep(Duration::from_secs(1)).await;
-
-    raft_1.leave().await;
-
-    sleep(Duration::from_secs(2)).await;
-
-    let raft_2 = rafts.get_mut(&2).unwrap();
-
-    wait_for_until_cluster_size_decrease(raft_2.clone(), 2).await;
-
-    let leader_id = raft_2.get_leader_id().await;
-
-    let timer = timeout(Duration::from_secs(5), async {
-        while leader_id == 0 {
-            sleep(Duration::from_secs(1)).await;
-        }
-    })
-    .await;
-
-    assert!(timer.is_ok(), "Actual leader_id: {}", leader_id);
-    assert!(
-        [2, 3].contains(&leader_id),
-        "Actual leader_id: {}",
-        leader_id
-    );
+    let mut current_node_cnt = raft_list.len();
+
+    let mut rafts = wait_until_rafts_ready(None, rx_raft, raft_list.len()).await;
+    let mut ex_term = 1;
+    for _ in 1..3 {
```

Comment:
Instead of performing it twice arbitrarily, how about passing the number of iterations as an argument?

Reply:
That'll be much vetter